### DISCUSSION TITLE
tool/onedrive/request

### DISCUSSION THREAD

**Discussion by giscus** [AlistGo/docs#175] [Announcements]
# tool/onedrive/request

import Onedrive from 

https://alist.nn.ci/tool/onedrive/request.html

**Comment by fourthwrite** [comment]
I’m getting an error from Microsoft. 

invalid_request: The provided value for the input parameter 'redirect_uri' is not valid. The expected value is a URI which matches a redirect URI registered for this client application.

**Reply by TTTPOB** [reply]
https://alist.nn.ci/tool/onedrive/callback url changed to this, this document seems not updated

**Reply by TTTPOB** [reply]
I mean, call back url in your azure application

**Comment by JsonSong89** [comment]
使用自己创建的client,刷新token报错:
我们无法完成你的请求
unauthorized_client: The client does not exist or is not enabled for consumers. If you are the application developer, configure a new application through the App Registrations in the Azure Portal at https://go.microsoft.com/fwlink/?linkid=2083908.
然后直接使用默认的0a2991a3-1674-4334-8561-671cc7349960,发现可以使用

**Reply by JsonSong89** [reply]
好吧,是我自己弄错了,client id 应该用应用注册那个页面的,我用了里面的机密id

**Comment by xiaojiedev** [comment]
是否可以将这个部分开源/出部署教程？
公共API回调地址可能会被微软拉黑，导致用户的订阅出问题，例子：https://qyi.io/archives/1042.html

**Reply by xiaojiedev** [reply]
或者像gonelist一样更好，部署URL就是回调地址。
https://gonelist-doc.cugxuan.cn/#/install?id=%e6%b3%a8%e5%86%8c%e5%ba%94%e7%94%a8%e7%a8%8b%e5%ba%8f

**Reply by xhofe** [reply]
文档站本身就是开源的：https://github.com/alist-org/docs/tree/main/docs/.vuepress/components/onedrive

**Comment by sH1222J** [comment]
invalid_request: The provided value for the input parameter 'redirect_uri' is not valid. The expected value is a URI which matches a redirect URI registered for this client application.

**Comment by rbforelle** [comment]
能否添加一个选项设置文件读写的权限？

根据这篇[微软的文档](https://learn.microsoft.com/zh-cn/onedrive/developer/rest-api/getting-started/graph-oauth?view=odsp-graph-online)中代码流相关的内容，作用域是可以选files.read保证文件只读的，但现在get refresh token按钮按了采用files.readwrite.all来提交request。所以如果有人想设置只读权限的，只要把点了ger refresh token跳出的网页地址中的`scope=offline_access+files.readwrite.all`改成`scope=offline_access+files.read.all`即可。地址形如：
```
https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id={client_id}&response_type=code&redirect_uri={redirect_uri}&scope=offline_access+files.read.all&state={state}
```
改了scope请求的权限就只有读取了，确定授权以后跳转回来就能正常得到refresh_token。在alist中实测有删除按钮但点了会显示Access Denied
